File: articles/social-cognitive-approaches-to-personality-goals-appraisals-and-self-regulation/python/analyze_social_cognitive_personality.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_social_cognitive_personality.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

PROCESSES = [
    "goal_activation",
    "threat_appraisal",
    "challenge_appraisal",
    "self_efficacy",
    "self_regulation",
    "emotional_arousal",
    "perceived_support",
]
OUTCOMES = ["prosocial_behavior", "avoidance_behavior", "task_persistence"]
DERIVED = ["appraisal_balance", "regulation_capacity", "approach_orientation", "avoidance_pressure"]

def ols(df, y, xs):
    X = np.column_stack([np.ones(len(df)), df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, df[y].to_numpy(float), rcond=None)
    return dict(zip(["intercept", *xs], map(float, beta)))

df = pd.read_csv(DATA)
required = {"person_id", "occasion", "situation_type", *PROCESSES, *OUTCOMES, *DERIVED}
missing = required - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

person_summary = (
    df.groupby("person_id")
    .agg(
        n_occasions=("occasion", "count"),
        goal_activation_mean=("goal_activation", "mean"),
        threat_appraisal_mean=("threat_appraisal", "mean"),
        challenge_appraisal_mean=("challenge_appraisal", "mean"),
        self_efficacy_mean=("self_efficacy", "mean"),
        self_regulation_mean=("self_regulation", "mean"),
        prosocial_behavior_mean=("prosocial_behavior", "mean"),
        avoidance_behavior_mean=("avoidance_behavior", "mean"),
        task_persistence_mean=("task_persistence", "mean"),
    )
    .reset_index()
)

situation_summary = (
    df.groupby("situation_type")
    .agg(
        n=("situation_type", "count"),
        goal_activation_mean=("goal_activation", "mean"),
        threat_appraisal_mean=("threat_appraisal", "mean"),
        challenge_appraisal_mean=("challenge_appraisal", "mean"),
        self_efficacy_mean=("self_efficacy", "mean"),
        self_regulation_mean=("self_regulation", "mean"),
        emotional_arousal_mean=("emotional_arousal", "mean"),
        perceived_support_mean=("perceived_support", "mean"),
        prosocial_behavior_mean=("prosocial_behavior", "mean"),
        avoidance_behavior_mean=("avoidance_behavior", "mean"),
        task_persistence_mean=("task_persistence", "mean"),
    )
    .reset_index()
)

if_then_summary = (
    df.groupby("situation_type")
    .agg(
        threat_appraisal_mean=("threat_appraisal", "mean"),
        challenge_appraisal_mean=("challenge_appraisal", "mean"),
        prosocial_behavior_mean=("prosocial_behavior", "mean"),
        avoidance_behavior_mean=("avoidance_behavior", "mean"),
        task_persistence_mean=("task_persistence", "mean"),
    )
    .reset_index()
)

person_summary.to_csv(OUT / "python_person_summary.csv", index=False)
situation_summary.to_csv(OUT / "python_situation_summary.csv", index=False)
if_then_summary.to_csv(OUT / "python_if_then_summary.csv", index=False)

corr_cols = PROCESSES + DERIVED + OUTCOMES
df[corr_cols].corr().to_csv(OUT / "python_social_cognitive_correlations.csv")

predictors = [
    "goal_activation",
    "threat_appraisal",
    "challenge_appraisal",
    "self_efficacy",
    "self_regulation",
    "emotional_arousal",
    "perceived_support",
]
models = {
    "prosocial_behavior": ols(df, "prosocial_behavior", predictors),
    "avoidance_behavior": ols(df, "avoidance_behavior", predictors),
    "task_persistence": ols(df, "task_persistence", predictors),
    "note": "Synthetic demonstration only; not personality testing, clinical assessment, educational placement, workplace screening, or individual ranking.",
}

(OUT / "python_social_cognitive_models.json").write_text(json.dumps(models, indent=2))
df.to_csv(OUT / "python_scored_social_cognitive_personality.csv", index=False)
print(f"Wrote outputs to {OUT}")

File: articles/social-cognitive-approaches-to-personality-goals-appraisals-and-self-regulation/r/analyze_social_cognitive_personality.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_social_cognitive_personality.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

processes <- c(
  "goal_activation",
  "threat_appraisal",
  "challenge_appraisal",
  "self_efficacy",
  "self_regulation",
  "emotional_arousal",
  "perceived_support"
)

outcomes <- c("prosocial_behavior", "avoidance_behavior", "task_persistence")
derived <- c("appraisal_balance", "regulation_capacity", "approach_orientation", "avoidance_pressure")

person_summary <- aggregate(
  data[, c(processes, outcomes)],
  by = list(person_id = data$person_id),
  FUN = mean
)
write.csv(person_summary, file.path(outputs, "r_person_summary.csv"), row.names = FALSE)

situation_summary <- aggregate(
  data[, c(processes, outcomes)],
  by = list(situation_type = data$situation_type),
  FUN = mean
)
write.csv(situation_summary, file.path(outputs, "r_situation_summary.csv"), row.names = FALSE)

corr_cols <- c(processes, derived, outcomes)
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_social_cognitive_correlations.csv"))

model_prosocial <- lm(
  prosocial_behavior ~ goal_activation + threat_appraisal + challenge_appraisal +
    self_efficacy + self_regulation + emotional_arousal + perceived_support,
  data = data
)

model_avoidance <- lm(
  avoidance_behavior ~ goal_activation + threat_appraisal + challenge_appraisal +
    self_efficacy + self_regulation + emotional_arousal + perceived_support,
  data = data
)

model_persistence <- lm(
  task_persistence ~ goal_activation + threat_appraisal + challenge_appraisal +
    self_efficacy + self_regulation + emotional_arousal + perceived_support,
  data = data
)

capture.output(summary(model_prosocial), file = file.path(outputs, "r_prosocial_behavior_model_summary.txt"))
capture.output(summary(model_avoidance), file = file.path(outputs, "r_avoidance_behavior_model_summary.txt"))
capture.output(summary(model_persistence), file = file.path(outputs, "r_task_persistence_model_summary.txt"))

write.csv(data, file.path(outputs, "r_scored_social_cognitive_personality.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")
